src/plugins/saved_objects/public/finder/saved_object_finder.tsx

```diff
@@ -155,7 +157,28 @@ class SavedObjectFinderUi extends React.Component<
       defaultSearchOperator: 'AND',
     });
 
-    resp.savedObjects = resp.savedObjects.filter((savedObject) => {
+    const getDataSource = async (id: string) => {
+      const client = this.props.savedObjects.client;
+      return await client.get<DataSourceAttributes>('data-source', id);
+    };
+
+    const savedObjects = await Promise.all(
+      resp.savedObjects.map(async (obj) => {
+        if (obj.type === 'index-pattern') {
+          const result = { ...obj };
+          result.attributes.title = await getIndexPatternTitle(
+            obj.attributes.title!,
+            obj.references,
+            getDataSource
+          );
+          return result;
```

Comment:
Prepending datasource to title requires fetching datasource name first, which replies on a saved object client to make a call. This should happen where there's a client to use. 
 - `saved object management definition` doesn't provide client , passing it to `magement -> getTitlte()` will change the interface that shard by other saved object. And to passing it for index-pattern only again need explicit check.
 - `index-pattern saved object` can't provide client, because unfortunately unlike `savedVis` that you referred to, there's no `SavedIndexPattern extends SavedObjectClass` exists. And it may not be worth it to create one only for prepending datasource to the title for display purpose. 

The basic idea is, the actual title saved for each index pattern(no matter which datasource it's created from) should be untouched to avoid blast radius, where title is used as a search field. We only change the index pattern title that being displayed, if an index pattern is created from a datasource.

Besides, there's an issue under discussion about having an OUI component that can be used to select datasource then index pattern when creating Visualzaiton/Dashboards/Discover. https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2841. We may need to do a refactor after that discussion is settled.